File: app/src/utils/aggregations.ts
import type { NutrientFruit, NutrientKey, FruitCategory } from '../types';
import { ALL_CATEGORIES } from './nutrition-meta';

export interface CategoryAverage {
  category: FruitCategory;
  count: number;
  averages: Record<NutrientKey, number | null>;
}

function avg(values: (number | null)[]): number | null {
  const valid = values.filter((v): v is number => v !== null);
  if (valid.length === 0) return null;
  return valid.reduce((sum, v) => sum + v, 0) / valid.length;
}

export function computeCategoryAverages(
  fruits: NutrientFruit[],
  nutrientKeys: NutrientKey[]
): CategoryAverage[] {
  const grouped = new Map<FruitCategory, NutrientFruit[]>();

  for (const cat of ALL_CATEGORIES) {
    grouped.set(cat as FruitCategory, []);
  }

  for (const fruit of fruits) {
    const group = grouped.get(fruit.category);
    if (group) group.push(fruit);
  }

  const result: CategoryAverage[] = [];

  for (const [category, catFruits] of grouped) {
    const averages: Record<string, number | null> = {};

    for (const key of nutrientKeys) {
      const values = catFruits.map((f) => f[key] as number | null);
      averages[key] = avg(values);
    }

    result.push({
      category,
      count: catFruits.length,
      averages: averages as Record<NutrientKey, number | null>,
    });
  }

  return result;
}

File: ios/Nutritionmaxx/Nutritionmaxx/Services/Aggregations.swift
import Foundation

struct CategoryAverage: Identifiable {
    let category: ItemCategory
    let count: Int
    let averages: [NutrientKey: Double?]

    var id: String { category.rawValue }
}

func computeCategoryAverages(
    foods: [FoodItem],
    nutrientKeys: [NutrientKey],
    categories: [ItemCategory]? = nil
) -> [CategoryAverage] {
    let cats = categories ?? ItemCategory.allCases
    var grouped: [ItemCategory: [FoodItem]] = [:]
    for cat in cats {
        grouped[cat] = []
    }

    for food in foods {
        grouped[food.category, default: []].append(food)
    }

    var result: [CategoryAverage] = []

    for cat in cats {
        let catFoods = grouped[cat] ?? []
        var averages: [NutrientKey: Double?] = [:]

        for key in nutrientKeys {
            let values = catFoods.compactMap { $0[key] }
            if values.isEmpty {
                averages[key] = nil
            } else {
                averages[key] = values.reduce(0, +) / Double(values.count)
            }
        }

        result.append(CategoryAverage(category: cat, count: catFoods.count, averages: averages))
    }

    return result.filter { $0.count > 0 }
}
